### PR TITLE
Configuring option parser

### PR DESCRIPTION
**Pull request by kiwiroy**
As an alternative to #20.

```perl
use Applify getopt_config => [qw(no_auto_help no_pass_through)];
```

### CODE REVIEW

File: t/options.t [hunk 1]
```diff
@@ -7,10 +7,19 @@ my $script = $app->_script;
...
 
-{
-  local $TODO = 'need to define config for Getopt::Long';
-  is_deeply($script->option_parser->{settings}, [qw(no_auto_help pass_through)], 'Getopt::Long has correct config');
-}
+getopt_long_config_ok($script->option_parser,
+  [qw(no_auto_help no_auto_version pass_through)],
+  'Getopt::Long has correct config');
+
+$app    =
```
Comment (jhthorsen) [inline t/options.t]: Is this file run through perltidy?

File: lib/Applify.pm [hunk 2]
```diff
@@ -758,6 +761,19 @@ a normalized matter. Example:
 Will export the functions listed under L</EXPORTED FUNCTIONS>. The functions
 will act on a L<Applify> object created by this method.
 
+Additionally, C<import> accepts the following named arguments with the following
+names:
+
+=over 4
```
Comment (jhthorsen) [inline lib/Applify.pm]: This should be `=over 2` for consistency.

File: lib/Applify.pm [hunk 3]
```diff
@@ -87,6 +87,9 @@ sub import {
 
   $self->{skip_subs} = {app => 1, option => 1, version => 1, documentation => 1, extends => 1, subcommand => 1};
 
+  $self->option_parser->configure(@{delete $args{getopt_config}})
+    if $args{getopt_config};
```
Comment (jhthorsen) [inline lib/Applify.pm]: I think we should merge this with the `option_parser()` attribute, and maybe make it read-only:

```
sub import {
  ...
  $args{getopt_config} ||= [qw(no_auto_help no_auto_version pass_through)];
  $self->option_parser->configure(@{delete $args{getopt_config}});
  ...
}
...
sub option_parser {
  return shift->{option_parser} ||= do { require Getopt::Long; Getopt::Long::Parser->new };
}
```

**Comment by jhthorsen** [comment]
I do have a completely different suggestion now: How about introducing hooks?

Something like:

```
hook around_run => sub { my ($script, $app) = @_; };
hook before_exit => sub { my ($script, $exit_code) = @_; };
hook before_options_parser => sub { my ($script, $argv) = @_; };
```

I think I would like that better, since that code will not run on compile time, and should also increase the flexibility.

**Comment by srchulo** [comment]
@jhthorsen I agree with the increased flexibility, that's nice. I liked the original syntax because it's easier to specify (less code), but that's just for my use case so I think it's valid that the more flexible option is better (and not much more code anyway).

What's the benefit of the code not running on compile time?

**Comment by jhthorsen** [comment]
> What's the benefit of the code not running on compile time?

I'm not sure. Just have a bad feeling about it...

**Pull request closed without merging**